File: backendNodeJS/src/controllers/newAccessToken.controller.js
import jwt from "jsonwebtoken";
import { response } from "../utils/response.util.js";
import { User } from "../models/user.model.js";
import mongoose from "mongoose";

const newAccessToken = async (req, res) => {
    const refreshToken = req.cookies.refreshToken;

    if (!refreshToken) return response(res, 401, "Refresh Token was not sent");

    try {
        const decoded = jwt.verify(refreshToken, process.env.JWT_KEY);
        const userid = decoded.id;
        const id = new mongoose.Types.ObjectId(userid);

        const user = await User.findOne({ _id: id });

        if (!user) return response(res, 401, "Unauthorized access was found");

        if (refreshToken != user.refreshToken)
            return response(res, 401, "Unauthorized access was found");

        const newAccessToken = jwt.sign(
            {
                id: user._id,
                username: user.username,
                email: user.email,
            },
            process.env.JWT_KEY,
            {
                expiresIn: "15m",
            }
        );

        const newRefreshToken = jwt.sign(
            {
                id: user._id,
            },
            process.env.JWT_KEY,
            {
                expiresIn: "15d",
            }
        );

        user.refreshToken = newRefreshToken;
        await user.save();

        return res
            .status(200)
            .cookie("accessToken", newAccessToken, {
                httpOnly: true,
                secure: true,
                sameSite: "None",
            })
            .cookie("refreshToken", newRefreshToken, {
                httpOnly: true,
                secure: true,
                sameSite: "None",
            })
            .json({
                message: "User authorized successfully",
                data: {
                    id: user._id,
                    username: user.username,
                    email: user.email,
                },
                error: "none",
            });
    } catch (error) {
        response(res, 401, "Unauthorized access was found");
    }
};

export { newAccessToken };

File: backendFlask/src/controllers/new_access_token_controller.py
from flask import request, jsonify
import jwt
from datetime import datetime, timedelta
from models.user import User
from utils.response_util import response
from config import JWT_SECRET_KEY


def new_access_token():
    refresh_token = request.cookies.get("refreshToken")
    if not refresh_token:
        return response(401, "Refresh Token was not sent")

    try:
        decoded = jwt.decode(refresh_token, JWT_SECRET_KEY, algorithms=["HS256"])
        user_id = decoded["id"]

        user = User.find_by_id(user_id)
        if not user or refresh_token != user["refreshToken"]:
            return response(401, "Unauthorized access was found")

        new_access_token = jwt.encode(
            {
                "id": str(user["_id"]),
                "username": user["username"],
                "email": user["email"],
                "exp": datetime.utcnow() + timedelta(minutes=15),
            },
            JWT_SECRET_KEY,
            algorithm="HS256",
        )

        new_refresh_token = jwt.encode(
            {"id": str(user["_id"]), "exp": datetime.utcnow() + timedelta(days=15)},
            JWT_SECRET_KEY,
            algorithm="HS256",
        )

        User.update_refresh_token(user["_id"], new_refresh_token)

        resp = jsonify(
            {
                "message": "User authorized successfully",
                "data": {
                    "id": str(user["_id"]),
                    "username": user["username"],
                    "email": user["email"],
                },
                "error": "none",
            }
        )
        resp.set_cookie(
            "accessToken", new_access_token, httponly=True, secure=True, samesite="None"
        )
        resp.set_cookie(
            "refreshToken",
            new_refresh_token,
            httponly=True,
            secure=True,
            samesite="None",
        )
        return resp, 200

    except jwt.ExpiredSignatureError:
        return response(401, "Refresh token has expired")
    except jwt.InvalidTokenError:
        return response(401, "Invalid refresh token")
